File: app/api/datasource.py
"""
数据源控制器，提供数据源相关的API接口
"""

from fastapi import APIRouter
from app.services.datasource.service import DatasourceService
from app.services.datasource.dto import DatasourceCreate, DatasourceUpdate
from app.utils.response import ResponseUtil, ApiResponse

router = APIRouter()


def format_datasource_data(data: dict) -> dict:
    """格式化数据源数据，移除ID中的横杠"""
    if data.get('id'):
        data['id'] = str(data['id']).replace('-', '')
    if data.get('category_id'):
        data['category_id'] = str(data['category_id']).replace('-', '')
    return data


@router.get("/types", response_model=ApiResponse)
def get_datasource_types():
    """
    获取支持的数据源类型列表
    
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    datasource_types = DatasourceService.get_datasource_types_with_config()
    return ResponseUtil.success(data=datasource_types, message="获取数据源类型列表成功")


@router.post("", response_model=ApiResponse)
def create_datasource(datasource: DatasourceCreate):
    """
    创建数据源
    
    Args:
        datasource: 数据源创建DTO
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    db_datasource = DatasourceService.create_datasource(datasource)
    return ResponseUtil.created(data=format_datasource_data(db_datasource.__data__.copy()), message="数据源创建成功")


@router.get("", response_model=ApiResponse)
def get_datasources(skip: int = 0, limit: int = 100, category_id: str = None, name: str = None, code: str = None, datasource_type: str = None):
    """
    获取数据源列表
    
    Args:
        skip: 跳过的记录数
        limit: 返回的最大记录数
        category_id: 分类ID（可选）
        name: 数据源名称（模糊查询）
        code: 数据源编码（模糊查询）
        datasource_type: 数据源类型
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    datasources, total = DatasourceService.get_datasources(skip, limit, category_id, name, code, datasource_type)
    return ResponseUtil.success(data={"data": datasources, "total": total}, message="获取数据源列表成功")


@router.post("/test_connection", response_model=ApiResponse)
def test_datasource_connection(data: dict):
    """
    测试数据源连接（用于新建数据源时）
    
    Args:
        data: 数据源测试数据
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    result = DatasourceService.test_connection_with_data(data)
    if result.get('success'):
        return ResponseUtil.success(data=result, message=result.get('message', '连接测试成功'))
    else:
        return ResponseUtil.error(code=400, message=result.get('message', '连接测试失败'), data=result)


@router.get("/{datasource_id}", response_model=ApiResponse)
def get_datasource(datasource_id: str):
    """
    获取单个数据源
    
    Args:
        datasource_id: 数据源ID
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    datasource = DatasourceService.get_datasource(datasource_id)
    if datasource is None:
        return ResponseUtil.not_found(message=f"数据源 {datasource_id} 不存在")
    return ResponseUtil.success(data=datasource, message="获取数据源成功")


@router.post("/{datasource_id}", response_model=ApiResponse)
def update_datasource(datasource_id: str, datasource: DatasourceUpdate):
    """
    更新数据源
    
    Args:
        datasource_id: 数据源ID
        datasource: 数据源更新DTO
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    db_datasource = DatasourceService.update_datasource(datasource_id, datasource)
    return ResponseUtil.success(data=format_datasource_data(db_datasource.__data__.copy()), message="数据源更新成功")


@router.post("/{datasource_id}/delete", response_model=ApiResponse)
def delete_datasource(datasource_id: str):
    """
    删除数据源
    
    Args:
        datasource_id: 数据源ID
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    db_datasource = DatasourceService.delete_datasource(datasource_id)
    return ResponseUtil.success(data=format_datasource_data(db_datasource.__data__.copy()), message="数据源删除成功")


@router.post("/{datasource_id}/test_connection", response_model=ApiResponse)
def test_datasource_connection_by_id(datasource_id: str):
    """
    测试数据源连接
    
    Args:
        datasource_id: 数据源ID
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    result = DatasourceService.test_connection(datasource_id)
    if result.get('success'):
        return ResponseUtil.success(data=result, message=result.get('message', '连接测试成功'))
    else:
        return ResponseUtil.error(code=400, message=result.get('message', '连接测试失败'), data=result)


@router.post("/{datasource_id}/query", response_model=ApiResponse)
def execute_datasource_query(datasource_id: str, query_data: dict):
    """
    执行数据源查询
    
    Args:
        datasource_id: 数据源ID
        query_data: 包含查询语句和参数的对象
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    query = query_data.get('query', '')
    params = query_data.get('params')
    result = DatasourceService.execute_query(datasource_id, query, params)
    if result.get('success'):
        return ResponseUtil.success(data=result, message=result.get('message', '查询执行成功'))
    else:
        return ResponseUtil.error(code=400, message=result.get('message', '查询执行失败'), data=result)


@router.get("/{datasource_id}/schema", response_model=ApiResponse)
def get_datasource_schema(datasource_id: str):
    """
    获取数据源Schema信息
    
    Args:
        datasource_id: 数据源ID
        
    Returns:
        ApiResponse: 统一格式的响应对象
    """
    result = DatasourceService.get_schema_info(datasource_id)
    if result.get('success'):
        return ResponseUtil.success(data=result, message=result.get('message', '获取Schema信息成功'))
    else:
        return ResponseUtil.error(code=400, message=result.get('message', '获取Schema信息失败'), data=result)

File: web/src/services/datasource.ts
/**
 * 数据源服务
 * 提供数据源相关的API调用
 */

import http from '../utils/request';

export interface DatasourceCategory {
  id: string;
  name: string;
  description?: string;
  is_default: boolean;
  parent_id?: string;
  sort_order: number;
  created_at?: string;
  updated_at?: string;
  children?: DatasourceCategory[];
}

export interface Datasource {
  id: string;
  code: string;
  name: string;
  type: string;
  category_id?: string;
  config?: Record<string, any>;
  status: boolean;
  created_at: string;
  updated_at?: string;
}

export interface DatasourceType {
  datasource_type: string;
  datasource_name: string;
  config_fields: any[];
  category?: string;
}

export const datasourceService = {
  /**
   * 获取数据源分类列表
   * @param parent_id - 父分类ID，为null时获取顶级分类
   */
  getCategories: async (parent_id?: string | null): Promise<DatasourceCategory[]> => {
    let params = [];
    if (parent_id !== undefined && parent_id !== null) {
      params.push(`parent_id=${parent_id}`);
    }
    const queryString = params.length > 0 ? `?${params.join('&')}` : '';
    return http.get<DatasourceCategory[]>(`/aicenter/v1/datasource_category${queryString}`);
  },

  /**
   * 获取分类树结构
   */
  getCategoryTree: async (): Promise<DatasourceCategory[]> => {
    return http.get<DatasourceCategory[]>('/aicenter/v1/datasource_category/tree');
  },

  /**
   * 创建分类
   * @param data - 分类数据
   */
  createCategory: async (data: Partial<DatasourceCategory>): Promise<DatasourceCategory> => {
    return http.post<DatasourceCategory>('/aicenter/v1/datasource_category', data);
  },

  /**
   * 更新分类
   * @param id - 分类ID
   * @param data - 分类数据
   */
  updateCategory: async (id: string, data: Partial<DatasourceCategory>): Promise<DatasourceCategory> => {
    return http.post<DatasourceCategory>(`/aicenter/v1/datasource_category/${id}`, data);
  },

  /**
   * 删除分类
   * @param id - 分类ID
   */
  deleteCategory: async (id: string): Promise<DatasourceCategory> => {
    return http.post<DatasourceCategory>(`/aicenter/v1/datasource_category/${id}/delete`);
  },

  /**
   * 更新分类排序
   * @param id - 分类ID
   * @param sort_order - 新的排序值
   */
  updateCategoryOrder: async (id: string, sort_order: number): Promise<DatasourceCategory> => {
    return http.post<DatasourceCategory>(`/aicenter/v1/datasource_category/${id}/order`, { sort_order });
  },

  /**
   * 获取支持的数据源类型列表
   */
  getDatasourceTypes: async (): Promise<DatasourceType[]> => {
    return http.get<DatasourceType[]>('/aicenter/v1/datasource/types');
  },

  /**
   * 获取数据源列表
   * @param categoryId - 分类ID（可选）
   * @param page - 页码，默认1
   * @param pageSize - 每页数量，默认12
   * @param name - 数据源名称（模糊查询）
   * @param code - 数据源编码（模糊查询）
   * @param datasourceType - 数据源类型
   */
  getDatasources: async (categoryId?: string, page: number = 1, pageSize: number = 12, name?: string, code?: string, datasourceType?: string): Promise<{ data: Datasource[], total: number }> => {
    let params = [];
    if (categoryId) params.push(`category_id=${categoryId}`);
    const skip = (page - 1) * pageSize;
    params.push(`skip=${skip}`);
    params.push(`limit=${pageSize}`);
    if (name) params.push(`name=${encodeURIComponent(name)}`);
    if (code) params.push(`code=${encodeURIComponent(code)}`);
    if (datasourceType) params.push(`datasource_type=${encodeURIComponent(datasourceType)}`);
    const queryString = params.length > 0 ? `?${params.join('&')}` : '';
    return http.get<{ data: Datasource[], total: number }>(`/aicenter/v1/datasource${queryString}`);
  },

  /**
   * 获取单个数据源
   * @param id - 数据源ID
   */
  getDatasource: async (id: string): Promise<Datasource> => {
    return http.get<Datasource>(`/aicenter/v1/datasource/${id}`);
  },

  /**
   * 创建数据源
   * @param data - 数据源数据
   */
  createDatasource: async (data: Partial<Datasource>): Promise<Datasource> => {
    return http.post<Datasource>('/aicenter/v1/datasource', data);
  },

  /**
   * 更新数据源
   * @param id - 数据源ID
   * @param data - 数据源数据
   */
  updateDatasource: async (id: string, data: Partial<Datasource>): Promise<Datasource> => {
    return http.post<Datasource>(`/aicenter/v1/datasource/${id}`, data);
  },

  /**
   * 删除数据源
   * @param id - 数据源ID
   */
  deleteDatasource: async (id: string): Promise<Datasource> => {
    return http.post<Datasource>(`/aicenter/v1/datasource/${id}/delete`);
  },

  /**
   * 测试数据源连接
   * @param id - 数据源ID
   */
  testConnection: async (id: string, data?: any): Promise<any> => {
    if (id === 'test' && data) {
      return http.post<any>('/aicenter/v1/datasource/test_connection', data);
    }
    return http.post<any>(`/aicenter/v1/datasource/${id}/test_connection`);
  },

  /**
   * 执行数据源查询
   * @param id - 数据源ID
   * @param query - 查询语句
   * @param params - 查询参数
   */
  executeQuery: async (id: string, query: string, params?: any): Promise<any> => {
    return http.post<any>(`/aicenter/v1/datasource/${id}/query`, { query, params });
  },

  /**
   * 获取数据源Schema信息
   * @param id - 数据源ID
   */
  getSchema: async (id: string): Promise<any> => {
    return http.get<any>(`/aicenter/v1/datasource/${id}/schema`);
  },

  /**
   * 列出文件（仅适用于文件存储类型数据源）
   * @param id - 数据源ID
   * @param bucket - Bucket名称（可选）
   * @param prefix - 文件前缀/目录（可选）
   * @param maxKeys - 最大返回数量
   * @param searchKeyword - 搜索关键词（可选）
   */
  listFiles: async (id: string, bucket?: string, prefix?: string, maxKeys: number = 100, searchKeyword?: string): Promise<any> => {
    let params = [`max_keys=${maxKeys}`];
    if (bucket) params.push(`bucket=${encodeURIComponent(bucket)}`);
    if (prefix) params.push(`prefix=${encodeURIComponent(prefix)}`);
    if (searchKeyword) params.push(`search_keyword=${encodeURIComponent(searchKeyword)}`);
    const queryString = params.length > 0 ? `?${params.join('&')}` : '';
    return http.get<any>(`/aicenter/v1/datasource/${id}/files${queryString}`);
  },

  /**
   * 下载文件（仅适用于文件存储类型数据源）
   * @param id - 数据源ID
   * @param bucket - Bucket名称（可选）
   * @param objectName - 对象名称/文件路径
   */
  downloadFile: async (id: string, bucket: string | undefined, objectName: string): Promise<any> => {
    return http.post<any>(`/aicenter/v1/datasource/${id}/files/download`, { bucket, object_name: objectName });
  },

  /**
   * 列出数据库表（仅适用于关系型数据库数据源）
   * @param id - 数据源ID
   * @param database - 数据库名称（可选）
   */
  listTables: async (id: string, database?: string): Promise<any> => {
    let params = [];
    if (database) params.push(`database=${encodeURIComponent(database)}`);
    const queryString = params.length > 0 ? `?${params.join('&')}` : '';
    return http.get<any>(`/aicenter/v1/datasource/${id}/tables${queryString}`);
  },

  /**
   * 获取表的字段信息（仅适用于关系型数据库数据源）
   * @param id - 数据源ID
   * @param tableName - 表名称
   * @param database - 数据库名称（可选）
   */
  getTableColumns: async (id: string, tableName: string, database?: string): Promise<any> => {
    let params = [];
    if (database) params.push(`database=${encodeURIComponent(database)}`);
    const queryString = params.length > 0 ? `?${params.join('&')}` : '';
    return http.get<any>(`/aicenter/v1/datasource/${id}/tables/${encodeURIComponent(tableName)}/columns${queryString}`);
  },
};
